docs/content/advance/configuration.md


# __NeoGPT Configuration__

This document outlines various configuration settings for NeoGPT. You can customize these settings to suit your needs.

!!! warning "Note"
    The default configurations are good enough for most use cases. You can skip this document if you don't want to customize the configurations. We don't recommend changing the configurations unless you know what you are doing.

!!! info "Note"
    The configuration file is located at `neogpt/config.py`. You can edit the file to change the configurations.

## Directories

- **Source Directory:** The directory where documents are stored.
  Default: `neogpt/documents`

- **Model Directory:** The directory for storing models from HuggingFace.
  Default: `neogpt/models`

- **Parent Database Directory:** The main database directory.
  Default: `neogpt/db`

- **Chroma Database Directory:** Directory for the Chroma database.
  Default: `neogpt/db/chroma`

- **FAISS Database Directory:** Directory for the FAISS database.
  Default: `neogpt/db/faiss`


## Memory and Model

- **Default Memory Key:** The default memory key to remember chat history.
  Default: `2`

- **Model Name:** The name of the GGUF model.
  Default: `TheBloke/Mistral-7B-Instruct-v0.1-GGUF`

- **Model File:** The model file to use.
  Default: `mistral-7b-instruct-v0.1.Q4_K_M.gguf`

- **Embedding Model:** The default embedding model.
  Default: `sentence-transformers/all-MiniLM-L12-v2`

## Threads and Device

- **Ingest Threads:** The number of threads for document ingestion.
  Default: `8`

- **Max Token Length:** The maximum token length for the model.
  Default: `8192`

- **Number of GPU Layers:** Number of layers for GPU compatibility.
  Default: `40`
  MPS: `1`

- **Device Type:** The device type (cpu, mps, cuda).
  Default: `cuda`

## Chroma Database Settings

- **Chroma Settings:** Anonymized telemetry and persistence settings.
    - Anonymized Telemetry: `False`
    - Is Persistent: `True`

## Reserved File Names

- List of reserved file names.
  Default: `["builder.url"]`

## Supported Document Extensions

- Supported document formats and their corresponding loaders.
  Default:

  - `.pdf`: PDFMinerLoader

  - `.txt`: TextLoader

  - `.csv`: CSVLoader
  - ... (other formats)

Read more about the supported document formats [here](../builder.md#supported-file-formats).


## URL Extensions

Supported URL patterns for ingestion.

  - `.youtube`: YoutubeLoader


## Query and Cost

- **Initial Query Cost:** Initial query cost.
  Default: `0`

- **Total Cost:** Total cost.
  Default: `0`

## Logging

- **Log Folder:** The folder for log files.
  Default: `logs`

- **Log File:** The log file name.
  Default: `logs/builder.log`

## Agent

- **Agent Name:** The name of the agent.
  Default: `NeoGPT`

- **Agent Thoughts:** List of agent thoughts.
  Default: `[]` (empty list)

- **QA feedback:** List of QA feedback.
  Default: `[]` (empty list)



!!! info "Update Note"
    You can update the configurations by using `export` command. For example, to update the `MODEL_NAME` configuration, you can use the following command:

    ```bash
      export MODEL_NAME="TheBloke/Mistral-7B-Instruct-v0.1-GGUF"
    ```


neogpt/config.py
import os

import torch
from chromadb.config import Settings
from colorama import init
from dotenv import load_dotenv
from langchain.chat_loaders.whatsapp import WhatsAppChatLoader
from langchain.document_loaders import (
    CSVLoader,
    JSONLoader,
    PDFMinerLoader,
    RecursiveUrlLoader,
    TextLoader,
    UnstructuredEmailLoader,
    UnstructuredEPubLoader,
    UnstructuredExcelLoader,
    UnstructuredHTMLLoader,
    UnstructuredMarkdownLoader,
    UnstructuredPowerPointLoader,
    UnstructuredTSVLoader,
    UnstructuredWordDocumentLoader,
    WebBaseLoader,
    YoutubeLoader,
)
from langchain.text_splitter import Language

# Load Environment Variables
load_dotenv()
# Initialize Colorama
init()

# Source Directory for Documents to Ingest
SOURCE_DIR = os.path.join(os.path.dirname(__file__), "documents")
# To store models from HuggingFace
MODEL_DIRECTORY = os.path.join(os.path.dirname(__file__), "models")
# PARENT DB DIRECTORY
PARENT_DB_DIRECTORY = os.path.join(os.path.dirname(__file__), "db")
# CHROMA DB DIRECTORY
CHROMA_PERSIST_DIRECTORY = os.path.join(PARENT_DB_DIRECTORY, "chroma")
# FAISS DB DIRECTORY
FAISS_PERSIST_DIRECTORY = os.path.join(PARENT_DB_DIRECTORY, "faiss")
# QDRANT PERSIST DIRECTORY
QDRANT_PERSIST_DIRECTORY = os.path.join(PARENT_DB_DIRECTORY, "qdrant")
# PINECONE DB DIRECTORY
PINECONE_PERSIST_DIRECTORY = os.path.join(PARENT_DB_DIRECTORY, "pinecone")
# WORKSPACE DIRECTORY
WORKSPACE_DIRECTORY = os.path.join(os.path.dirname(__file__), "workspace")

#path of json file to pass the commandline arguments to ui
UI_ARGS_PATH=os.path.join(os.path.dirname(__file__), "ui_args.json")

# DEFAULT MEMORY KEY FOR CONVERSATION MEMORY (DEFAULT IS 2)
DEFAULT_MEMORY_KEY = 2

# GGUF MODELS (Recommended , Default and Fast)
MODEL_NAME = os.getenv("MODEL_NAME", "TheBloke/Mistral-7B-Instruct-v0.1-GGUF")
MODEL_FILE = os.getenv("MODEL_FILE", "mistral-7b-instruct-v0.1.Q4_K_M.gguf")

# MISTRAL MODEL LITE
# MODEL_NAME = "TheBloke/MistralLite-7B-GGUF"
# MODEL_FILE = "mistrallite.Q4_K_M.gguf"

# GPTQ MODELS (Quantized Models)
# MODEL_NAME = "TheBloke/Tinyllama-2-1b-miniguanaco-GPTQ"
# MODEL_FILE = "model.safetensors"

# HUGGING FACE MODEL (Not recommended for low RAM systems)
# MODEL_NAME = "microsoft/phi-1_5"
# MODEL_FILE = None

# DEFAULT EMBEDDING MODEL
EMBEDDING_MODEL = "sentence-transformers/all-MiniLM-L12-v2"
# EMBEDDING MODEL CONFIG
INGEST_THREADS = 8

# MODEL CONFIG
MAX_TOKEN_LENGTH = 8192  # 8192 is the max for Mistral-7B
N_GPU_LAYERS = 40

# PYTORCH DEVICE COMPATIBILITY
if torch.cuda.is_available():
    DEVICE_TYPE = "cuda"
elif torch.backends.mps.is_available():
    DEVICE_TYPE = "mps"
else:
    DEVICE_TYPE = "cpu"

# CHROMA DB SETTINGS
CHROMA_SETTINGS = Settings(anonymized_telemetry=False, is_persistent=True)

# PINECONE SETTINGS
EMBEDDING_DIMENSION = ""
INDEX_NAME = ""

# Reserved File Names
RESERVED_FILE_NAMES = ["builder.url"]

# List of file supported for ingest
DOCUMENT_EXTENSION = {
    ".pdf": PDFMinerLoader,
    ".txt": TextLoader,
    ".csv": CSVLoader,
    ".html": UnstructuredHTMLLoader,
    ".tsv": UnstructuredTSVLoader,
    ".eml": UnstructuredEmailLoader,
    ".epub": UnstructuredEPubLoader,
    ".xls": UnstructuredExcelLoader,
    ".xlsx": UnstructuredExcelLoader,
    ".pptx": UnstructuredPowerPointLoader,
    ".ppt": UnstructuredPowerPointLoader,
    ".docx": UnstructuredWordDocumentLoader,
    ".doc": UnstructuredWordDocumentLoader,
    ".md": UnstructuredMarkdownLoader,
    ".json": JSONLoader,
    # ".py": TextLoader,
}

# List of URL patterns supported for ingest
URL_EXTENSION = {
    ".youtube": YoutubeLoader,
    "recursive": RecursiveUrlLoader,
    "normal": WebBaseLoader,
}

# List of all Social Chat and their loaders
SOCIAL_CHAT_EXTENSION = {
    r"^(chat_|_chat|whatsapp_|whatsapp_chat|whatsapp_chat_|whatsapp_)": WhatsAppChatLoader
}


# List of all file extensions for programming languages and their parsers
CODE_EXTENSION = {
    ".cpp": Language.CPP,
    ".go": Language.GO,
    ".java": Language.JAVA,
    ".kt": Language.KOTLIN,
    ".js": Language.JS,
    ".ts": Language.TS,
    ".php": Language.PHP,
    ".proto": Language.PROTO,
    ".py": Language.PYTHON,
    ".rst": Language.RST,
    ".ruby": Language.RUBY,
    ".rs": Language.RUST,
    ".scala": Language.SCALA,
    ".swift": Language.SWIFT,
    ".markdown": Language.MARKDOWN,
    ".latex": Language.LATEX,
    ".html": Language.HTML,
    ".sol": Language.SOL,
    ".cs": Language.CSHARP,
    ".cobol": Language.COBOL,
}


# Initial Query Cost and Total Cost
QUERY_COST = 0
TOTAL_COST = 0

# LOG CONFIG
LOG_FOLDER = os.path.join(os.path.dirname(__file__), "logs")
# BUILDER LOG
BUILDER_LOG_FILE = os.path.join(LOG_FOLDER, "builder.log")
# NEOGPT LOG
NEOGPT_LOG_FILE = os.path.join(LOG_FOLDER, "neogpt.log")


# # Export Config

# def export_config():
#     config = {
#         "neogpt" : {
#             "VERSION": "0.1.0-alpha",
#             "ENV": "development",
#             "PERSONA": "default",
#             "UI" : False,
#             "MODEL_TYPE": "mistral",
#             "EXPORT_DATE": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),

#         },
#         "model" : {
#             "MODEL_NAME": MODEL_NAME,
#             "MODEL_FILE": MODEL_FILE,
#             "EMBEDDING_MODEL": EMBEDDING_MODEL,
#         },
#         "database" : {
#             "PARENT_DB_DIRECTORY": os.path.basename(PARENT_DB_DIRECTORY),
#         },
#         "directories" : {
#             "SOURCE_DIR": os.path.basename(SOURCE_DIR),
#             "WORKSPACE_DIRECTORY": os.path.basename(WORKSPACE_DIRECTORY),
#         },
#         "memory" : {
#             "DEFAULT_MEMORY_KEY": DEFAULT_MEMORY_KEY,
#         },

#     }

#     with open("settings/settings.yaml", "w") as file:
#         yaml.dump(config, file,sort_keys=False)


# AGENT CONFIG
PROJECT_COST = 0
AGENT_THOUGHTS = []
QA_ENGINEER_FEEDBACK = ""
CURRENT_WORKING_AGENT = ["NeoGPT"]
